certora/specs/Gates.spec

```
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: GPL-2.0-or-later
+// Copyright (c) 2025 Morpho Association
+
+using ERC20Standard as ERC20;
+using MorphoMarketV1Adapter as MorphoMarketV1Adapter;
+using MorphoVaultV1Adapter as MorphoVaultV1Adapter;
+
+methods {
+    function multicall(bytes[]) external => NONDET DELETE;
+
+    function asset() external returns address envfree;
+    function balanceOf(address) external returns uint256 envfree;
+    function canReceiveShares(address) external returns bool envfree;
+    function canSendShares(address) external returns bool envfree;
+    function canSendAssets(address) external returns bool envfree;
+    function canReceiveAssets(address) external returns bool envfree;
+    function ERC20.totalSupply() external returns uint256 envfree;
+
+    function _.canSendShares(address user) external => ghostStatusCanSendShares[user] expect bool;
+    function _.canReceiveShares(address user) external => ghostStatusCanReceiveShares[user] expect bool;
+    function _.canSendAssets(address user) external =>  ghostStatusCanSendAssets[user] expect bool;
+    function _.canReceiveAssets(address user) external => ghostStatusCanReceiveAssets[user] expect bool;
+
+    function _.allocate(bytes data, uint256 assets, bytes4, address) external => DISPATCHER(true);
+    function _.deallocate(bytes data, uint256 assets, bytes4, address) external => DISPATCHER(true);
+    function _.realizeLoss(bytes, bytes4, address) external => DISPATCHER(true);
+
+    function _.supply(MorphoMarketV1Adapter.MarketParams, uint256, uint256, address, bytes) external => NONDET;
+    function _.withdraw(MorphoMarketV1Adapter.MarketParams, uint256, uint256, address, address) external => NONDET;
+    function _.deposit(uint256, address) external => NONDET;
+    function _.withdraw(uint256, address, address) external => NONDET;
+    function _.interestPerSecond(uint256, uint256) external => NONDET;
+}
+
+persistent ghost mapping(address => bool) ghostStatusCanSendShares {
+    init_state axiom forall address user. ghostStatusCanSendShares[user] == false;
+}
+
+persistent ghost mapping(address => bool) ghostStatusCanReceiveShares {
+    init_state axiom forall address user. ghostStatusCanReceiveShares[user] == false;
+}
+
+persistent ghost mapping(address => bool) ghostStatusCanSendAssets {
+    init_state axiom forall address user. ghostStatusCanSendAssets[user] == false;
+}
+
+persistent ghost mapping(address => bool) ghostStatusCanReceiveAssets {
+    init_state axiom forall address user. ghostStatusCanReceiveAssets[user] == false;
+}
+
+persistent ghost mapping(address => bool) ghostBalanceChangeAllowed {
+    init_state axiom forall address user. ghostBalanceChangeAllowed[user] == true;
+}
+
+hook Sstore ERC20.balanceOf[KEY address user] uint256 newBalance (uint256 oldBalance) {
+    if (!canReceiveAssets(user)) {
+        ghostBalanceChangeAllowed[user] = oldBalance >= newBalance && ghostBalanceChangeAllowed[user];
```

```suggestion
        assert oldBalance >= newBalance;
```
I find this clearer, because it removes one ghost variable, and also you have to understand how the hook works anyway (you can't just look at the spec)